### PR TITLE
feat: Update image list on version

### PR DESCRIPTION
**Pull request by milindk8**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes [#AB1617](https://dev.azure.com/DevtronLabs/Devtron/_workitems/edit/1617/)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
All the tests have been done manually

- [x] By changing the image from select in all three pages i.e cluster list, node list, node details
- [x] By hovering & selecting images & testing version through command line

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/common/helpers/Helpers.tsx [hunk 1]
```diff
@@ -1014,16 +1014,33 @@ export const createGroupedItemsByKey = (arr: any[], key: string) => {
...
         return {
-            label: customObjToPickLabel ? customObjToPickLabel[ele]?.label : labelKey ? ele[labelKey] : ele,
-            value: valueKey ? ele[valueKey] : ele,
+            label: customObjToPickLabel ? ele.name : ele,
+            value: customObjToPickLabel ? ele.image : ele,
         }
     })
 }
+
+export const clusterImageTippy = (nodeImageList,selectedImage): string => {
```
Comment (vivek-devtron) [inline src/components/common/helpers/Helpers.tsx:1040]: Please rename this method to clusterImageDescription

File: src/components/common/helpers/Helpers.tsx [hunk 2]
```diff
@@ -1014,16 +1014,33 @@ export const createGroupedItemsByKey = (arr: any[], key: string) => {
...
 }
 
+export const regexImageList = (imageList,serverVersion) => {
+
+    let nodeImageList = imageList.find((imageObj) => {
+        let regex = new RegExp(imageObj.groupRegex)
+        return regex.test(serverVersion)
+    })
+    
+    return nodeImageList.imageList
```
Comment (vivek-devtron) [inline src/components/common/helpers/Helpers.tsx:1024]: it seems incorrect, nodeImageList will be array type and that should be returned directly

File: src/components/ClusterNodes/types.ts [hunk 3]
```diff
@@ -137,7 +137,7 @@ export interface ClusterTerminalType {
     clusterName?: string
     nodeList: string[]
     closeTerminal?: () => void
-    clusterImageList: string[]
+    clusterImageList: any[]
```
Comment (vivek-devtron) [inline src/components/ClusterNodes/types.ts:140]: Provide type for imageList

File: src/components/common/helpers/Helpers.tsx (hunk 1 above)
Comment (milindk8) [reply src/components/common/helpers/Helpers.tsx]: done

File: src/components/common/helpers/Helpers.tsx (hunk 2 above)
Comment (milindk8) [reply src/components/common/helpers/Helpers.tsx]: done

File: src/components/ClusterNodes/types.ts (hunk 3 above)
Comment (milindk8) [reply src/components/ClusterNodes/types.ts]: done

File: src/components/ClusterNodes/ClusterNodeContainer.tsx [hunk 4]
```diff
@@ -8,17 +8,18 @@ import { getHostURLConfiguration } from '../../services/service'
...
     const [isSuperAdmin, setSuperAdmin] = useState<boolean>()
     const [namespaceDefaultList, setNameSpaceList] = useState<string[]>()
     
     useEffect(() => {
         Promise.all([getHostURLConfiguration('DEFAULT_TERMINAL_IMAGE_LIST'),getUserRole(),clusterNamespaceList()]).then(([hostUrlConfig, userRole, namespaceList]) => {
             if(hostUrlConfig.result) {
-                const imageValue: string = hostUrlConfig.result.value || ''
-                setImageList(imageValue.split(','))
+                const imageValue: string = hostUrlConfig.result.value
+                setImageList(JSON.parse(imageValue))
```
Comment (sohel-rp) [inline src/components/ClusterNodes/ClusterNodeContainer.tsx:22]: Please wrap this inside try...catch

Comment (milindk8) [reply src/components/ClusterNodes/ClusterNodeContainer.tsx]: done

**Review by vivek-devtron — APPROVED** [review]

File: src/components/common/helpers/Helpers.tsx [hunk 5]
```diff
@@ -1014,16 +1015,38 @@ export const createGroupedItemsByKey = (arr: any[], key: string) => {
...
-            label: customObjToPickLabel ? customObjToPickLabel[ele]?.label : labelKey ? ele[labelKey] : ele,
-            value: valueKey ? ele[valueKey] : ele,
+            label: customLabel ? ele[customLabel] : ele,
+            value: customValue ? ele[customValue] : ele,
         }
     })
 }
+
+export const clusterImageDescription = (nodeImageList,selectedImage): string => {
+    let nodeImageObj = nodeImageList.find((obj) => {
```
Comment (sohel-rp) [inline src/components/common/helpers/Helpers.tsx:1048]: make it const

File: src/components/common/helpers/Helpers.tsx [hunk 6]
```diff
@@ -1014,16 +1015,38 @@ export const createGroupedItemsByKey = (arr: any[], key: string) => {
     }, {})
 }
 
+export const filterImageList = (imageList, serverVersion): ClusterImageList[] => {
+    let nodeImageList = imageList.find((imageObj) => {
+        let regex = new RegExp(imageObj.groupRegex)
```
Comment (sohel-rp) [inline src/components/common/helpers/Helpers.tsx:1020]: same here

File: src/components/common/helpers/Helpers.tsx [hunk 7]
```diff
@@ -1014,16 +1015,38 @@ export const createGroupedItemsByKey = (arr: any[], key: string) => {
     }, {})
 }
 
+export const filterImageList = (imageList, serverVersion): ClusterImageList[] => {
```
Comment (sohel-rp) [inline src/components/common/helpers/Helpers.tsx:1018]: missing args type

File: src/components/common/helpers/Helpers.tsx [hunk 8]
```diff
@@ -1014,16 +1015,38 @@ export const createGroupedItemsByKey = (arr: any[], key: string) => {
...
         return {
-            label: customObjToPickLabel ? customObjToPickLabel[ele]?.label : labelKey ? ele[labelKey] : ele,
-            value: valueKey ? ele[valueKey] : ele,
+            label: customLabel ? ele[customLabel] : ele,
+            value: customValue ? ele[customValue] : ele,
         }
     })
 }
+
+export const clusterImageDescription = (nodeImageList,selectedImage): string => {
```
Comment (sohel-rp) [inline src/components/common/helpers/Helpers.tsx:1047]: same here

File: src/components/ClusterNodes/NodeList.tsx [hunk 9]
```diff
@@ -69,6 +70,7 @@ export default function NodeList({ imageList, isSuperAdmin, namespaceList }: Clu
     const [showTerminal, setTerminal] = useState<boolean>(false)
     const nodeList = filteredFlattenNodeList.map((node) => node['name'])
     const clusterName: string = filteredFlattenNodeList[0]?.['clusterName'] || ''
+    const [nodeImageList, setNodeImageList] = useState([])
```
Comment (sohel-rp) [inline src/components/ClusterNodes/NodeList.tsx:73]: provide type

File: src/components/ClusterNodes/NodeDetails.tsx [hunk 10]
```diff
@@ -67,6 +68,7 @@ export default function NodeDetails({ imageList, isSuperAdmin, namespaceList }:
     const [lastDataSync, setLastDataSync] = useState(false)
     const [isShowWarning, setIsShowWarning] = useState(false)
     const [patchData, setPatchData] = useState<jsonpatch.Operation[]>(null)
+    const [nodeImageList, setNodeImageList] = useState([])
```
Comment (sohel-rp) [inline src/components/ClusterNodes/NodeDetails.tsx:71]: same here

Comment (sohel-rp) [inline src/components/ClusterNodes/NodeDetails.tsx:71]: same here

File: src/components/ClusterNodes/ClusterTerminal.tsx [hunk 11]
```diff
@@ -347,6 +349,20 @@ export default function ClusterTerminal({
         )
     }
 
+    const imageOptionComponent = (props) => {
+        let tippyText = clusterImageDescription(clusterImageList, props.data.value)
```
Comment (sohel-rp) [inline src/components/ClusterNodes/ClusterTerminal.tsx:353]: same here

Comment (milindk8) [reply src/components/ClusterNodes/ClusterTerminal.tsx]: done

File: src/components/ClusterNodes/NodeDetails.tsx (hunk 10 above)
Comment (milindk8) [reply src/components/ClusterNodes/NodeDetails.tsx]: done

Comment (milindk8) [reply src/components/ClusterNodes/NodeDetails.tsx]: done

File: src/components/ClusterNodes/NodeList.tsx (hunk 9 above)
Comment (milindk8) [reply src/components/ClusterNodes/NodeList.tsx]: done

File: src/components/common/helpers/Helpers.tsx (hunk 8 above)
Comment (milindk8) [reply src/components/common/helpers/Helpers.tsx]: done

File: src/components/common/helpers/Helpers.tsx (hunk 7 above)
Comment (milindk8) [reply src/components/common/helpers/Helpers.tsx]: done

File: src/components/common/helpers/Helpers.tsx (hunk 6 above)
Comment (milindk8) [reply src/components/common/helpers/Helpers.tsx]: done

File: src/components/common/helpers/Helpers.tsx (hunk 5 above)
Comment (milindk8) [reply src/components/common/helpers/Helpers.tsx]: done

**Review by sohel-rp — APPROVED** [review]

**Pull request merged**